src/main/java/htsjdk/samtools/GATKReadSparkCodec.java

```diff
@@ -0,0 +1,196 @@
+                "; quals length: " + alignment.getBaseQualities().length);
+            }
+            this.binaryCodec.writeString(alignment.getReadName(), false, true);
+            final int[] binaryCigar = BinaryCigarCodec.encode(alignment.getCigar());
+            for (final int cigarElement : binaryCigar) {
+                // Assumption that this will fit into an integer, despite the fact
+                // that it is specced as a uint.
+                this.binaryCodec.writeInt(cigarElement);
+            }
+            this.binaryCodec.writeBytes(SAMUtils.bytesToCompressedBases(alignment.getReadBases()));
+            byte[] qualities = alignment.getBaseQualities();
+            if (qualities.length == 0) {
+                qualities = new byte[alignment.getReadLength()];
+                Arrays.fill(qualities, (byte) 0xFF);
+            }
+            this.binaryCodec.writeBytes(qualities);
+            SAMBinaryTagAndValue attribute = alignment.getBinaryAttributes();
+            while (attribute != null) {
+                this.binaryTagCodec.writeTag(attribute.tag, attribute.value, attribute.isUnsignedArray());
+                attribute = attribute.getNext();
+            }
+        }
+    }
+
+    /**
+     * Read the next record from the input stream and convert into a java object.
+     *
+     * @return null if no more records.  Should throw exception if EOF is encountered in the middle of
+     *         a record.
+     */
+    public SAMRecord decode() {
+        int recordLength = 0;
+        try {
+            recordLength = this.binaryCodec.readInt();
+        }
+        catch (RuntimeEOFException e) {
+            return null;
+        }
+
+        if (recordLength < BAMFileConstants.FIXED_BLOCK_SIZE) {
+            throw new SAMFormatException("Invalid record length: " + recordLength);
+        }
+        
+        final int referenceID = this.binaryCodec.readInt();
+        final int coordinate = this.binaryCodec.readInt() + 1;
+        final short readNameLength = this.binaryCodec.readUByte();
+        final short mappingQuality = this.binaryCodec.readUByte();
+        final int bin = this.binaryCodec.readUShort();
+        final int cigarLen = this.binaryCodec.readUShort();
+        final int flags = this.binaryCodec.readUShort();
+        final int readLen = this.binaryCodec.readInt();
+        final int mateReferenceID = this.binaryCodec.readInt();
+        final int mateCoordinate = this.binaryCodec.readInt() + 1;
+        final int insertSize = this.binaryCodec.readInt();
+        final byte[] restOfRecord = new byte[recordLength - BAMFileConstants.FIXED_BLOCK_SIZE];
+        this.binaryCodec.readBytes(restOfRecord);
+        final BAMRecord ret = this.samRecordFactory.createBAMRecord(
+                header, referenceID, coordinate, readNameLength, mappingQuality,
+                bin, cigarLen, flags, readLen, mateReferenceID, mateCoordinate, insertSize, restOfRecord);
+        ret.setHeader(header); 
```

Comment:
Why are you restoring a header here? Shouldn't we just keep the reads headerless?